protocol/x/affiliates/keeper/keeper.go

```diff
@@ -41,3 +49,154 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	affiliateReferredVolumePrefixStore := prefix.NewStore(ctx.KVStore(k.storeKey), []byte(types.ReferredVolumeKeyPrefix))
+	if !affiliateReferredVolumePrefixStore.Has([]byte(affiliateAddr)) {
+		return dtypes.NewInt(0), false
+	}
+	var referredVolume dtypes.SerializableInt
+	err := referredVolume.Unmarshal(affiliateReferredVolumePrefixStore.Get([]byte(affiliateAddr)))
+	if err != nil {
+		return dtypes.NewInt(0), false
+	}
+	return referredVolume, true
+}
+
+func (k Keeper) GetAllAffiliateTiers(ctx sdk.Context) (types.AffiliateTiers, error) {
+	store := ctx.KVStore(k.storeKey)
+	affiliateTiersBytes := store.Get([]byte(types.AffiliateTiersKey))
+
+	var affiliateTiers types.AffiliateTiers
+	if affiliateTiersBytes == nil {
+		return affiliateTiers, errors.New("Affiliate tiers not found")
+	}
+	err := k.cdc.Unmarshal(affiliateTiersBytes, &affiliateTiers)
+	if err != nil {
+		return affiliateTiers, err
+	}
+
+	return affiliateTiers, nil
+}
+
+func (k Keeper) GetTakerFeeShare(
+	ctx sdk.Context,
+	address string,
+) (
+	affiliateAddress string,
+	feeSharePpm uint32,
+	exists bool,
+	err error,
+) {
+	affiliateAddress, exists = k.GetReferredBy(ctx, address)
+	if !exists {
+		return "", 0, false, nil
+	}
+	_, feeSharePpm, err = k.GetTierForAffiliate(ctx, affiliateAddress)
+	if err != nil {
+		return "", 0, false, err
+	}
+	return affiliateAddress, feeSharePpm, true, nil
+}
+
+// Assumes that the affiliate tiers are sorted by level in ascending order.
+func (k Keeper) GetTierForAffiliate(
+	ctx sdk.Context,
+	affiliateAddr string,
+) (
+	tierLevel uint32,
+	feeSharePpm uint32,
+	err error) {
+	affiliateTiers, err := k.GetAllAffiliateTiers(ctx)
+	if err != nil {
+		return 0, 0, err
+	}
+	numTiers := uint32(len(affiliateTiers.GetTiers()))
+	maxTierLevel := numTiers - 1
+	currentTier := uint32(0)
+	referredVolume, exists := k.GetReferredVolume(ctx, affiliateAddr)
+
+	if !exists {
+		// If referred volume is not found, set it to 0.
+		referredVolume = dtypes.NewInt(0)
+	}
+
+	for _, tier := range affiliateTiers.GetTiers() {
+		if referredVolume.BigInt().Int64() >= int64(tier.ReqReferredVolume) {
+			currentTier = tier.GetLevel()
+		}
+	}
+
+	if currentTier == maxTierLevel {
+		return currentTier, affiliateTiers.GetTiers()[currentTier].TakerFeeSharePpm, nil
+	}
+	numCoinsStaked := k.statsKeeper.GetStakedAmount(ctx, affiliateAddr)
+	for _, tier := range affiliateTiers.GetTiers() {
+		if numCoinsStaked.Cmp(big.NewInt(
+			int64(affiliateTiers.GetTiers()[currentTier].ReqStakedWholeCoins))) >= 0 &&
+			tier.GetLevel() > currentTier {
+			currentTier = tier.GetLevel()
+		}
+	}
+	return currentTier, affiliateTiers.GetTiers()[currentTier].TakerFeeSharePpm, nil
+}
+
+func (k Keeper) UpdateAffiliateTiers(ctx sdk.Context, affiliateTiers types.AffiliateTiers) error {
+	numTiers := uint32(len(affiliateTiers.GetTiers()))
+	for i := uint32(0); i < numTiers; i++ {
+		if affiliateTiers.GetTiers()[i].GetLevel() != i {
```

Comment:
Hmm how about this: remove the `Level` field in `AffiliateTier` object. The index of `affiliateTiers.Tiers[index]` inherently indicates the tier. This way we make the object more lightweight to serde, and avoid sorted check here

(Pls update proto message and spec accordingly, with comment that `index` maps to tier level

Reply:
Also no need to return error after this change